### DISCUSSION TITLE
Research frontend e2e testing with Playwright

### DISCUSSION THREAD

**Discussion by dhochbaum-dcp** [NYCPlanning/ae-zoning-map-poc#53] [General]
Based upon issue #52 
Branch with what I have so far: https://github.com/NYCPlanning/ae-zoning-map-poc/compare/52-research-e2e-with-playwright

Initial thoughts:

- Pros
  - Tools are intuitive and easy to use
  - Both the [test recorder](https://playwright.dev/docs/codegen-intro) and the [debugging UI](https://playwright.dev/docs/running-tests#debug-tests-in-ui-mode) are the best I have ever used
  - Mock API is simple
- Cons
  - Checking if a class exists is not that intuitive:
  `expect(page.locator(".example")).not.toHaveClass(["example"])`
  - Simulating an API error request seems to require recording a HAR file, although the one I created can be re-used

**Comment by TylerMatteo** [comment]
Good initial analysis @dhochbaum-dcp. I'll leave some comments for specific changes on the PR but wanted to leave some general thoughts here so others can chime in:

### To mock or not to mock?
I appreciate that David's PR shows what testing looks like with and without mocking the API. For anyone else with experience writing front end e2e tests, I'd love to hear if you're used to seeing these tests with or without mocked APIs, or both. Playwright's docs clearly suggest mocking them is a valid use case with the tooling they provide for that, but also [hint at](https://playwright.dev/docs/best-practices#testing-with-a-database) testing against staging environments when your FE is hitting APIs we also control. I can certainly see the value in either approach. Maybe it would make sense to test against staging environments from CI? Fwiw, I think our existing apps with e2e suites such as Applicant Portal test to mock API responses.

### How to mock APIs
I think using `har` files is a really interesting feature of Playwright and something to consider, but I owe @dhochbaum-dcp some direction as to how we might use Kubb to mock API calls based on OpenAPI specs. Maybe you could connect with @TangoYankee and @pratishta about this? They've been working with Kubb's `faker` plugin for backend testing, and we might be able to use the [MSW](https://www.kubb.dev/plugins/swagger-msw/) plugin in a similar way on the frontend.

**Reply by TangoYankee** [reply]
Historically, I've only used workflows based on mocked api requests, using [`fetch-mock`](https://www.npmjs.com/package/fetch-mock) and vanilla [`msw`](https://www.npmjs.com/package/msw). For the zoning-api, we did briefly explore Kubb's [`swagger/faker`](https://www.kubb.dev/plugins/swagger-faker/#kubb-swagger-faker) for backend testing. We opted not use it for the backend because we actually needed the repository faked, which the openapi docs (and subsequently the `swagger/faker` module) are blind to. The frontend may be a better use case, as we do need to fake the actual response from the server. Though, buyer beware, we ran into some issues faking objects when their schemas involved regular expressions - and a lot of our schemas involve regular expressions. 

> To mock or not to mock?

I lean towards a mock-first philosophy. Primarily, this is to give deference to the api "spec"/documentation. The mocks can be constructed around how the api is supposed to be constructed, solidifying the docs as the definitive source of truth. The API implementation may deviate from the documentation. If we are testing, and subsequently developing, around a flawed implementation there are two possible outcomes. In one scenario, we may not notice the deviation and will release a deviant API. This will frustrate anyone trying to use the API based on the documentation- both internally and externally. Alternatively, we may notice the deviation and choose to fix it. However, this means front-end development will be blocked until the API deviation is corrected in the back-end.

Other benefits derive from testing based on mocks: 
1. We do not need to wait for an implementation of the back-end to begin developing the front-end.  
2. We get more deterministic tests.
    - We have more control over the data in the mock, making responses more predictable. For example, a `200` test scenario that requests a `bbl` will know that the endpoint will always return a `200` because it controls what data is in the mock. This guarantee does not exist when hitting against a live environment that has its own data updates.
    - We can more easily test for server errors- faking them easily with a mock
4. We reduce traffic against the staging environment

**Reply by dhochbaum-dcp** [reply]
I think for tests to have value, they need to:
- Tell you whether or not the site is working
- Be updated to reflect changes to the site
- Not be ignored

When considering how to move forward, I think we should review our current strategy.  It took me roughly 2 minutes to come up with this list of active repos where we are actively ignoring failing tests and pushing changes to production:
- [Zola](https://github.com/NYCPlanning/labs-zola/pulls?q=is%3Apr)
- [Applicant Portal](https://github.com/NYCPlanning/labs-applicant-portal/pulls?q=is%3Apr)
- [Applicant Maps](https://github.com/NYCPlanning/labs-applicantmaps/pulls?q=is%3Apr)
- [Applicant Maps API](https://github.com/NYCPlanning/labs-applicantmaps-api/pulls?q=is%3Apr)
- [Layers API](https://github.com/NYCPlanning/labs-layers-api/pulls?q=is%3Apr)
- [Search API](https://github.com/NYCPlanning/labs-search-api/pulls?q=is%3Apr)
- [CEQR](https://github.com/NYCPlanning/ceqr-app/pulls?q=is%3Apr)

Mocking an API call is more work, results in more code, and ultimately provides a less reliable answer of whether the site is actually working as intended.  If we want to get value out of tests, I think we should keep them as simple as possible for now.

**Reply by TylerMatteo** [reply]
@dhochbaum-dcp  I take your point about ignoring tests - that's certainly not something we want to accept going forward. As for mocking, it's really matter of what we want the scope of a given test to be. I think a good place for us to start is to focusing on testing front end functionality from end to end from the perspective of user-facing functionality, assuming concerns external to the front end such as APIs are working as expected. Given that context, "end to end" refers to the use case (e.g. "I can visit the site, type in a bbl, click search, and see the tax lot details"), not end to end in a "vertical" sense where one end the front end and the other is the back end. Isolating these tests to the front end by mocking APIs based on their respective OpenAPI specifications mean that, when they do fail, we already know the problem is on the front end.

After thinking about whether or not we want to mock APIs a little more, the path forward I'm recommending is to start with tests that mock them, preferably leveraging the API's OpenAPI specification. We can revisit testing against live staging APIs, particularly in a workflow where those tests are run from CI environments, which would most likely mean GH Actions in our case.


Now for something completely different...

> Checking if a class exists is not that intuitive:
expect(page.locator(".example")).not.toHaveClass(["example"])

I'm curious if you find their whole page locator and expect API unintuitive or if you particularly had trouble with checking for DOM elements by clas sname? The reason I'm asking is that the writing tests that rely on the existence of elements with a given class name is actually something Playwright's best practices warns against, and is something I've seen discouraged in other documentation, such as that of React Testing Library. Instead, you want to write tests that test based on how a real user would actually interact with the site. Playwright's docs refer to this as "[user-visible behavior](https://playwright.dev/docs/best-practices#test-user-visible-behavior)". Avoiding the use of CSS classes in selector's is especially discouraged for several reasons such as:
* They are less reflective of how a user really uses the application, compared to, for example, selecting an element based on the text it contains
* They aren't unique - several elements could have the same class name, which can lead to flaky tests as other, seemingly unrelated functionality is added to the application
* They are less "semantic" than other methods of selecting DOM elements. That is to say, they don't tell us what an element is _for_ in the way something like `role="button"` does. Playwright's docs also steer towards using [those sorts of attributes for selection](https://playwright.dev/docs/best-practices#prefer-user-facing-attributes-to-xpath-or-css-selectors) and this approach is helpful when selecting by text is impossible or impractical.

**Reply by dhochbaum-dcp** [reply]
In this instance, I struggled to find the best method to confirm that the Tax Lot Details pane was not appearing.  In this instance, I added an ID to the pane, so it probably makes more sense to use `await expect(page.locator("#tax-lot-details")).not.toBeVisible();`
Is there another method that works better for this particular application?

**Reply by TylerMatteo** [reply]
In keeping with the spirit of testing the UI based on how real users interact with it, I don't think we need to test that the actual detail pane is visible. Your tests that check for the contents of the pane will accomplish that. For testing that certain bits of text appear, you should prefer _selecting by text_ rather than `id`. The text is "user facing" and using it means you don't need to add `id`s everywhere just for the sake of testing. So
```
await expect(page.locator("#tax-lot-details")).toContainText(
    "Tax Lot: BBL1000477501",
  );
```
can become
```
await expect(
      page.getByText("Tax Lot: BBL1000477501", { exact: true }), // exact flag is maybe unnecessary but doesn't hurt
    ).toBeVisible();
```

Playwright's docs [list](https://playwright.dev/docs/locators#quick-guide) some preferred built-in locator methods that you should default to. Selecting by `id` is probably preferable to selecting by `class` but you should pretty much never have to use either method. Selecting by text is generally the best approach  for inert elements like `span`s and `p`s. Using things like `getByRole` in conjunction with `name`[ like you do in your branch](https://github.com/NYCPlanning/ae-zoning-map-poc/pull/56/files#diff-0d390b6fa21eccc29f07f83632410abd29136fdcdbdb808983c4e6a44ccecf2eR33) works great for things like `button`s. If you really find yourself needing an escape, you may be testing for something that doesn't need a test, such as the existence of containers like the details pane. If you _really really_ need an escape hatch, [adding test ids ](https://playwright.dev/docs/locators#locate-by-test-id)is preferable to using classes or ids.
